src/librustc/middle/ty/item_path.rs

```diff
@@ -0,0 +1,317 @@
+
+        if !in_self_mod && !in_trait_mod {
+            // If the impl is not co-located with either self-type or
+            // trait-type, then fallback to a format that identifies
+            // the module more clearly.
+            self.push_item_path(buffer, parent_def_id);
+            if let Some(trait_ref) = impl_trait_ref {
+                buffer.push(&format!("<impl {} for {}>", trait_ref, self_ty));
+            } else {
+                buffer.push(&format!("<impl {}>", self_ty));
+            }
+            return;
+        }
+
+        // Otherwise, try to give a good form that would be valid language
+        // syntax. Preferably using associated item notation.
+
+        if let Some(trait_ref) = impl_trait_ref {
+            // Trait impls.
+            buffer.push(&format!("<{} as {}>",
+                                 self_ty,
+                                 trait_ref));
+            return;
+        }
+
+        // Inherent impls. Try to print `Foo::bar` for an inherent
+        // impl on `Foo`, but fallback to `<Foo>::bar` if self-type is
+        // anything other than a simple path.
+        match self_ty.sty {
+            ty::TyStruct(adt_def, substs) |
+            ty::TyEnum(adt_def, substs) => {
+                if substs.types.is_empty() { // ignore regions
+                    self.push_item_path(buffer, adt_def.did);
+                } else {
+                    buffer.push(&format!("<{}>", self_ty));
+                }
+            }
+
+            ty::TyBool |
+            ty::TyChar |
+            ty::TyInt(_) |
+            ty::TyUint(_) |
+            ty::TyFloat(_) |
+            ty::TyStr => {
+                buffer.push(&format!("{}", self_ty));
+            }
+
+            _ => {
+                buffer.push(&format!("<{}>", self_ty));
+            }
+        }
+    }
+
+    fn push_impl_path_fallback<T>(&self,
+                                  buffer: &mut T,
+                                  impl_def_id: DefId)
+        where T: ItemPathBuffer
+    {
+        // If no type info is available, fall back to
+        // pretty printing some span information. This should
+        // only occur very early in the compiler pipeline.
+        let parent_def_id = self.parent_def_id(impl_def_id).unwrap();
+        self.push_item_path(buffer, parent_def_id);
+        let node_id = self.map.as_local_node_id(impl_def_id).unwrap();
+        let item = self.map.expect_item(node_id);
+        let span_str = self.sess.codemap().span_to_string(item.span);
+        buffer.push(&format!("<impl at {}>", span_str));
+    }
+
+    /// As a heuristic, when we see an impl, if we see that the
+    /// 'self-type' is a type defined in the same module as the impl,
+    /// we can omit including the path to the impl itself. This
+    /// function tries to find a "characteristic def-id" for a
+    /// type. It's just a heuristic so it makes some questionable
+    /// decisions and we may want to adjust it later.
```

Comment:
On Tue, Mar 22, 2016 at 05:54:48AM -0700, arielb1 wrote:
> I am not sure that this is better than treating all impls as local. For example, `structural_impls` would use the ugly way - that's it, `middle::structural_impls::<impl TypeFoldable for Vec<T>>::fold_ty` rather than `<Vec<T> as TypeFoldable>::fold_ty`.

Yeah, I don't know what's best. One could also imagine some heuristic
based on "submodules". But I guess one could argue that for
`structural_impls` it'd be useful to be told that the impl is in that
module, rather than the module for `Vec`. I'm totally amenable to
follow-up PRs that tweak these heuristics -- probably at some point we
would want to RFC this, but I guess that's a long way in the future.